### PR TITLE
Added a scheduled task to prune old unused ips

### PR DESCRIPTION
**Pull request by solarkennedy**
---

I guess I didn't ask before starting, but is simply removing old entries in this table all this needs to do?

I don't know that much about postgres transactions and consistency. Are there any locking concerns that I should be aware of? (with other things that might be trying to get stuff from this table. No right? We are using a real db here).

I didn't test this on any live infra (only SELECT to validate the query), as it is destructive. Would you like me to do more with regards to testing?

### CODE REVIEW

**Comment by solarkennedy** [comment]
@sargun Ok I think I got the sql the way you want. I still made it select by vpc because the tasks are split out per-vpc anyway. It makes the sql more complex, but it makes it easier to reason about what each task is operating over?

Or maybe we don't need to fan out the tasks per vpc at all?

File: vpc/service/prune_unused_ips.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	ctx, span := trace.StartSpan(ctx, "pruneLastUsedIPAddresses")
+	defer span.End()
+
+	logger.G(ctx).WithFields(map[string]interface{}{
+		"vpcID": vpcID.String,
+	}).Info("Beginning purge of last used ip addresses")
+
+	result, err := tx.ExecContext(ctx,
```
Comment (sargun) [inline vpc/service/prune_unused_ips.go:70]: Is there a reason to have this scheduled task run on a per-vpc basis, versus using the nilitem and running on scheduled task periodically and doing the delete with a slightly more complex SQL query?

Comment (solarkennedy) [reply vpc/service/prune_unused_ips.go:70]: I don't know. I was mostly going off of your original advice to make the tasks fan out on a per-vpc basis.

File: vpc/service/prune_unused_ips.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+		if err != nil {
+			return nil, err
+		}
+		ret = append(ret, &vid)
+	}
+	_ = tx.Commit()
+	return ret, nil
+}
+
+type vpcID struct {
```
Comment (sargun) [inline vpc/service/prune_unused_ips.go:43]: Any reason this isn't just type vpcId string?

File: vpc/service/prune_unused_ips.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+func (v *vpcID) String() string {
+	return fmt.Sprintf(v.vpcID)
+}
+
+func (vpcService *vpcService) pruneLastUsedIPAddresses(ctx context.Context, vpcID keyedItem, tx *sql.Tx) (retErr error) {
+
+	//  Each Titus task requires an IPv4 address. Each VPC has 3 x /16s for Titus to operate in.
+	// That means that the maximum number of "interesting" IPv4 rows per table are 3*(2**16) = 196608
+	howManyToKeep := 200000
```
Comment (sargun) [inline vpc/service/prune_unused_ips.go:59]: this should probably be a const somewhere

File: vpc/service/prune_unused_ips.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+				last_seen,
+				RANK() OVER (PARTITION BY family(ip_address)
+							 ORDER BY last_seen DESC) AS rank
+		FROM ip_last_used_v3 WHERE AND vpc_id = '$2;)
+	    DELETE
+	    FROM ip_last_used_v3
+	    WHERE id IN
+		  (SELECT id
+		   FROM addresses_by_family_in_vpc
+		   WHERE rank > $1 AND last_seen > now() - INTERVAL '30 DAY')
```
Comment (sargun) [inline vpc/service/prune_unused_ips.go:82]: I think you can probably crank this down to something like 1 day.

File: vpc/service/prune_unused_ips.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+		"vpcID": vpcID.String,
+	}).Info("Beginning purge of last used ip addresses")
+
+	result, err := tx.ExecContext(ctx,
+		`WITH addresses_by_family_in_vpc AS
+		(SELECT id,
+				last_seen,
+				RANK() OVER (PARTITION BY family(ip_address)
+							 ORDER BY last_seen DESC) AS rank
+		FROM ip_last_used_v3 WHERE AND vpc_id = '$2;)
```
Comment (sargun) [inline vpc/service/prune_unused_ips.go:76]: no seimicolon

File: vpc/service/prune_unused_ips.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+	defer cancel()
+	ctx, span := trace.StartSpan(ctx, "pruneLastUsedIPAddresses")
+	defer span.End()
+
+	logger.G(ctx).WithFields(map[string]interface{}{
+		"vpcID": vpcID.String,
+	}).Info("Beginning purge of last used ip addresses")
+
+	result, err := tx.ExecContext(ctx,
+		`WITH addresses_by_family_in_vpc AS
```
Comment (sargun) [inline vpc/service/prune_unused_ips.go:71]: btw, I find https://sqlformat.org/ helpful

File: vpc/service/prune_unused_ips.go (hunk 5 above)
Comment (sargun) [inline vpc/service/prune_unused_ips.go:76]: is that quoting correct?

Comment (sargun) [inline vpc/service/prune_unused_ips.go:76]: It should be something like:
```
WITH addresses_by_family AS
  (SELECT id,
          last_seen,
          RANK() OVER (PARTITION BY family(ip_address)                       
                       ORDER BY last_seen DESC) AS rank
   FROM ip_last_used_v3 
WHERE vpc_id = $1 AND last_seen != TIMESTAMP 'EPOCH'
)
DELETE
FROM ip_last_used_v3
WHERE id IN
    (SELECT id
     FROM addresses_by_family
     WHERE rank > $2 AND last_seen > now() - INTERVAL '1 DAY');
```

I think?

**Review by sargun — CHANGES_REQUESTED** [review]

File: vpc/service/prune_unused_ips.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+		return err
+	}
+
+	rows, err := result.RowsAffected()
+	if err != nil {
+		err = errors.Wrap(err, "Failed to load the result")
+		return err
+	}
+
+	logger.G(ctx).WithFields(map[string]interface{}{
```
Comment (sargun) [inline vpc/service/prune_unused_ips.go:98]: If rows > 0; you may want to add an analyze statement here.

Comment (solarkennedy) [reply vpc/service/prune_unused_ips.go:98]: Why though? Wouldn't it be pretty normal to remove rows?

Comment (sargun) [reply vpc/service/prune_unused_ips.go:98]: Eh, I guess we can look at it later.

File: vpc/service/prune_unused_ips.go (hunk 1 above)
Comment (solarkennedy) [reply vpc/service/prune_unused_ips.go:70]: I'm going with nilitem

**Comment by solarkennedy** [comment]
This has some updates from last night and now doesn't do per-vpc. Ready for second round of review.

File: vpc/service/prune_unused_ips.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+				RANK() OVER (PARTITION BY family(ip_address), vpc_id
+							 ORDER BY last_seen DESC) AS rank
+		    FROM ip_last_used_v3
+		    WHERE last_seen != TIMESTAMP 'EPOCH')
+	    DELETE
+	    FROM ip_last_used_v3
+	    WHERE id IN
+		    (SELECT id
+		    FROM addresses_by_family
+		    WHERE rank > $1 AND last_seen > now() - INTERVAL '1 DAY')`,
```
Comment (sargun) [inline vpc/service/prune_unused_ips.go:41]: shouldn't this be:
`last_seen < now() - INTERVAL '1 DAY'`?

File: vpc/service/prune_unused_ips.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+		howManyUnusedIPsToKeep,
+	)
+	if err != nil {
+		err = errors.Wrap(err, "Failed to delete unused ip adddress")
+		return err
+	}
+
+	rows, err := result.RowsAffected()
+	if err != nil {
+		err = errors.Wrap(err, "Failed to load the result")
```
Comment (sargun) [inline vpc/service/prune_unused_ips.go:51]: Please user tracehelpers to set the trace status on error.

**Review by sargun — CHANGES_REQUESTED** [review]

File: vpc/service/prune_unused_ips.go (hunk 8 above)
Comment (solarkennedy) [reply vpc/service/prune_unused_ips.go:41]: Whoops! I updated your jira comment but didn't update the code... :(

File: vpc/service/prune_unused_ips.go [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+		return err
+	}
+
+	rows, err := result.RowsAffected()
+	if err != nil {
+		err = errors.Wrap(err, "Failed to load the result from the DELETE query to ip_last_used_v3s")
+		tracehelpers.SetStatus(err, span)
+		return err
+	}
+	logger.G(ctx).Infof("Deleted %d ip addresses (left the oldest %d)", rows, howManyUnusedIPsToKeep)
```
Comment (sargun) [inline vpc/service/prune_unused_ips.go:54]: Please use withfields here.

Comment (sargun) [reply vpc/service/prune_unused_ips.go:54]: or WithField() on the # or rows.

**Review by sargun — APPROVED** [review]
One small nit, and then we're good to go!

File: vpc/service/prune_unused_ips.go (hunk 10 above)
Comment (solarkennedy) [reply vpc/service/prune_unused_ips.go]: added

**Review by sargun — APPROVED** [review]

**Pull request merged**